service/history/workflow/cache/cache.go

```diff
@@ -86,10 +87,16 @@ type (
 	CacheImpl struct {
 		cache.Cache
 
+		logger                    log.Logger
 		nonUserContextLockTimeout time.Duration
 	}
+	cacheItem struct {
+		ctx       workflow.Context
+		logger    log.Logger
+		finalizer *common.Finalizer
+	}
```

Comment:
The cache value is wrapped with this new struct to provide access to the finalizer etc. It is hidden from the caller.